Utilities/Mergers/src/Merger.cxx

```diff
@@ -0,0 +1,284 @@
+
+#include <Framework/CompletionPolicyHelpers.h>
+#include <Framework/TimesliceIndex.h>
+#include <Framework/CallbackService.h>
+
+#include <TObjArray.h>
+#include <TH1.h>
+#include <TH2.h>
+#include <TH3.h>
+#include <THn.h>
+#include <TTree.h>
+#include <THnSparse.h>
+
+using namespace o2::framework;
+using namespace std::chrono;
+
+namespace o2
+{
+namespace mergers
+{
+
+Merger::Merger(MergerConfig config, header::DataHeader::SubSpecificationType subSpec)
+  : mConfig(config),
+    mSubSpec(subSpec),
+    mCache(config.ownershipMode.value == OwnershipMode::Full)
+{
+}
+
+Merger::~Merger()
+{
+}
+
+void Merger::init(framework::InitContext& ictx)
+{
+  if (mConfig.publicationDecision.value == PublicationDecision::EachNSeconds) {
+    // Register a device callback which creates timeslice in the TimesliceIndex
+    // each N seconds, so it can serve as timer input.
+    ictx.services().get<CallbackService>().set(CallbackService::Id::ClockTick, prepareTimerCallback(ictx));
+  }
+}
+
+void Merger::run(framework::ProcessingContext& ctx)
+{
+  if (mCache.uninitialized()) {
+    mCache.init(ctx.inputs());
+  }
+
+  mCache.cacheInputRecord(ctx.inputs());
+
+  if (shouldMergeCache(ctx)) {
+
+    mergeCache();
+
+    cleanCacheAfterMerging();
+  }
+
+  if (shouldPublish(ctx)) {
+
+    publish(ctx.outputs());
+
+    cleanCacheAfterPublishing();
+
+    if (mConfig.timespan.value == Timespan::LastDifference) {
+      mMergedObjects.reset();
+    }
+  }
+}
+
+std::function<void()> Merger::prepareTimerCallback(InitContext& ictx) const
+{
+  return [&timesliceIndex = ictx.services().get<TimesliceIndex>(),
+          timeLast = std::make_shared<steady_clock::time_point>(steady_clock::now()),
+          periodMs = this->mConfig.publicationDecision.param * 1000,
+          timesliceID = uint64_t(0) ]() mutable
+  {
+    auto timeNow = steady_clock::now();
+
+    if (duration_cast<milliseconds>(timeNow - *timeLast).count() > periodMs) {
+
+      data_matcher::VariableContext context;
+      context.put(data_matcher::ContextUpdate{ 0, timesliceID });
+      context.commit();
+
+      timesliceIndex.replaceLRUWith(context);
+
+      timesliceID++;
+      *timeLast = timeNow;
+    }
+  };
+}
+
+void Merger::cleanCacheAfterMerging()
+{
+  mCache.setAllMerged();
+}
+
+void Merger::cleanCacheAfterPublishing()
+{
+  if (mConfig.timespan.value == Timespan::LastDifference || mConfig.ownershipMode.value == OwnershipMode::Integral) {
+    mCache.clear();
+  }
+
+  if (mConfig.ownershipMode.value == OwnershipMode::Full) {
+    mCache.setAllMerged(false);
+  }
+  mCache.setAllUpdated(false);
+}
+
+void Merger::mergeCache()
+{
+  switch (mConfig.mergingMode.value) {
+    case MergingMode::Binwise:
+    case MergingMode::UseInterfaceFunction: {
+
+      size_t i = 0;
+      if (!mMergedObjects) {
+        for (; i < mCache.size(); i++) {
+          if (!mCache[i].deque.empty()) {
+            mMergedObjects.reset(mCache[i].deque[0].obj->Clone());
```

Comment:
why Clone()? why not just move the object as a first instance? Calls to Clone() are very expensive, also on memory (they go through the root streamer system)

Reply:
Moving the object would break one corner case.
Suppose that you merge objects from n inputs. If each of them appears regularly, then everything is fine. However, if one of producers fails to deliver the new object, we would have to either:
- wait until we receive it,
- merge without that missing object,
- use the last object that we received on that input.
I followed the third option, since it seems to produce most accurate results and would not be blocked by having just one input missing. It is possible to configure these mergers to merge and publish when certain % of inputs is updated, not necessarily 100%.
By cloning the object I keep the original intact, so it can be possibly reused in such case.

Reply:
optimizing for a corner case this early on will make maintenance very difficult. Anyway - my suggestion is as always - only send deltas and keep the merged state in the merger only - other cases are in my experience unmaintainable and slow in an online environment. We do have to keep in mind that this code needs to run synchronously with data taking so we need to trust is blindly.

Reply:
In the case of merging deltas, Clone() will be called only once, when the Merger has not receive anything yet. Afterwards, the merged object will be always present, so there will be no need to Clone() anything again. Calling it only once in the lifetime of the process should not be such a big deal.

I completely agree with you, that this code should not be overblown. On the other hand, this is a part of my future PhD thesis and I wanted to perform more thorough study of Mergers to describe and compare different approaches. I am not really sure what to propose as a solution - maybe I could split it into two versions, first simple and fine-tuned and second, more general. Still this would require from my side maintaining both.

Reply:
agree, calling Clone() once is not an issue - is you restart your process each time a starts, that will not necessarily be the case in the future - otherwise you pull in upstream library problems (ROOT is known to e.g. leak from time to time) each time you reconfigure your process.

As for the idea of splitting the code - it actually is a good one, there should, IMHO, be a clear split between production code and "playground" code.... and in the case of the former it should "do one thing and do it well" to cite the classics :)